frontend/src/routes/(authenticated)/project/[project_code]/BulkAddProjectMembers.svelte

```diff
@@ -0,0 +1,110 @@
+<script lang="ts">
+  import { BadgeButton, MemberBadge } from '$lib/components/Badges';
+  import { DialogResponse, FormModal } from '$lib/components/modals';
+  import { Input, ProjectRoleSelect, TextArea, passwordFormRules } from '$lib/forms';
+  import { ChangeProjectDescriptionDocument, ProjectRole } from '$lib/gql/types';
+  import t from '$lib/i18n';
+  import { z } from 'zod';
+  import { _bulkAddProjectMembers } from './+page';
+  import { hash } from '$lib/util/hash';
+  import { AdminContent } from '$lib/layout';
+  import { createEventDispatcher } from 'svelte';
+  import Icon from '$lib/icons/Icon.svelte';
+
+  enum BulkAddSteps {
+    Add,
+    Results,
+  }
+
+  let currentStep = BulkAddSteps.Add;
+
+  const dispatch = createEventDispatcher();
+
+  export let projectId: string;
+  const schema = z.object({
+    usernamesText: z.string().min(1, $t('register.name_missing')),
+    password: passwordFormRules($t),
+  });
+
+  let formModal: FormModal<typeof schema>;
+  $: form = formModal?.form();
+
+  let createdCount: number | undefined = undefined;
+  let usernameConflicts: string[] = [];
+
+  const usernameRe = /^[a-zA-Z0-9_]+$/;
+
+  function validateUsernames(usernames: string[]): boolean {
+    return usernames.every(s => usernameRe.test(s));
+  }
+
+  async function openModal(): Promise<void> {
+    const { response, formState } = await formModal.open(async () => {
+      const passwordHash = await hash($form.password);
+      const usernames = $form.usernamesText.split('\n').map(s => s.trim());
+      if (!validateUsernames(usernames)) {
+        return $t('project_page.bulk_add_members.usernames_alphanum_only');
+      }
+      const { error, data } = await _bulkAddProjectMembers({
+        projectId,
+        passwordHash,
+        usernames,
+        role: ProjectRole.Editor, // Managers not allowed to have shared passwords
```

Comment:
But they can become a manager if they get an email, verify it and (later) meet our password strength requirements?

Reply:
Yes. Once the password strength requirements get merged, bulk-added users will automatically have a password strength of 0 so they'll be automatically forced to change their password before they can become a project manager (because managers will be required to have a minimum password strength of 1 or 2, I'm leaning towards 2). And the verified email check that I just added (in a commit that isn't pushed yet) will take care of the email verification.

Reply:
The commit I mentioned is pushed now: e1391d6498280e77fd10d35730508154d351e23b.